src/merkle/tiered_smt/values.rs

```diff
@@ -0,0 +1,587 @@
+            } else {
+                let end_prefix = (parent_index.value() + 1) << (MAX_DEPTH - tier);
+                let mut iter = self.range(start_prefix..end_prefix);
+                iter.next().filter(|_| iter.next().is_none())
+            };
+
+            if let Some((key, value)) = sibling {
+                return Some((key, value, parent_index));
+            }
+        }
+
+        None
+    }
+
+    /// Returns an iterator over all key-value pairs in this store.
+    pub fn iter(&self) -> impl Iterator<Item = &(RpoDigest, Word)> {
+        self.values.iter().flat_map(|(_, entry)| entry.iter())
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts the specified key-value pair into this store and returns the value previously
+    /// associated with the specified key.
+    ///
+    /// If no value was previously associated with the specified key, None is returned.
+    pub fn insert(&mut self, key: RpoDigest, value: Word) -> Option<Word> {
+        let prefix = get_key_prefix(&key);
+        match self.values.entry(prefix) {
+            Entry::Occupied(mut entry) => entry.get_mut().insert(key, value),
+            Entry::Vacant(entry) => {
+                entry.insert(StoreEntry::new(key, value));
+                None
+            }
+        }
+    }
+
+    /// Removes the key-value pair for the specified key from this store and returns the value
+    /// associated with this key.
+    ///
+    /// If no value was associated with the specified key, None is returned.
+    pub fn remove(&mut self, key: &RpoDigest) -> Option<Word> {
+        let prefix = get_key_prefix(key);
+        match self.values.entry(prefix) {
+            Entry::Occupied(mut entry) => {
+                let (value, remove_entry) = entry.get_mut().remove(key);
+                if remove_entry {
+                    entry.remove_entry();
+                }
+                value
+            }
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    // HELPER METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an iterator over all key-value pairs contained in this store such that the most
+    /// significant 64 bits of the key lay within the specified bounds.
+    ///
+    /// The order of iteration is from the smallest to the largest key.
+    fn range<R: RangeBounds<u64>>(&self, bounds: R) -> impl Iterator<Item = &(RpoDigest, Word)> {
+        self.values.range(bounds).flat_map(|(_, entry)| entry.iter())
+    }
+}
+
+// VALUE NODE
+// ================================================================================================
+
+/// An entry in the [ValueStore].
+///
+/// An entry can contain either a single key-value pair or a vector of key-value pairs sorted by
+/// key.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum StoreEntry {
```

Comment:
We could, but the reason I went with the enum approach is to make it more efficient. Specifically, the dominant use case (by far) is that each leaf will have only a single value. So, my thinking was that we'd end up with a bunch of vector each containing just a single value.

This should result in performance penalty because:
1. We'd need to allocate these vectors on the heap.
2. It adds an extra level of indirection)
3. It would consume more memory (vector pointer and size + value vs. value + 1 byte for enum tag).

But I haven't benchmarked an alternative approach to know for sure.